Fancyauth.APIUtil/FancyPluginBase.cs

```diff
@@ -0,0 +1,436 @@
+        private readonly ConcurrentDictionary<int, ImmutableList<GlobalListener<IUser>>> SingleUserListeners = new ConcurrentDictionary<int, ImmutableList<GlobalListener<IUser>>>();
+        private readonly ConcurrentDictionary<int, ImmutableList<ActiveListener<IUser>>> UserListeners = new ConcurrentDictionary<int, ImmutableList<ActiveListener<IUser>>>();
+
+        private readonly ConcurrentBag<GlobalListener<IChannel>> GlobalChannelListeners = new ConcurrentBag<GlobalListener<IChannel>>();
+        private readonly ConcurrentDictionary<int, ImmutableList<GlobalListener<IChannel>>> SingleChannelListeners = new ConcurrentDictionary<int, ImmutableList<GlobalListener<IChannel>>>();
+        private readonly ConcurrentDictionary<int, ImmutableList<ActiveListener<IChannel>>> ChannelListeners = new ConcurrentDictionary<int, ImmutableList<ActiveListener<IChannel>>>();
+       
+        /// <summary>.
+        /// Init this instance. It fetches all Users and Channels from the Server and buffers them.
+        /// That way only Δ must be applied which saving resources.
+        /// </summary>
+        public async sealed override Task Init()
+        {
+            bool success = true;
+            foreach (var user in await Server.GetOnlineUsers())
+            {
+                success &= UserListeners.TryAdd(user.SessionId, ImmutableList<ActiveListener<IUser>>.Empty);
+            }
+            foreach (var channel in await Server.GetChannels())
+            {
+                success &= ChannelListeners.TryAdd(channel.ChannelId, ImmutableList<ActiveListener<IChannel>>.Empty);
+            }
+
+            if (!success)
+            {
+                throw new Exception("Init of FancyPluginBase failed due to non-empty listener-lists or duplicte IDs. " +
+                    "This should actually not happen - we are so screwed!!!");
+            }
+        }
+        
+        /// <summary>
+        /// Adds a Global Listener for an IReadModifyWrite-Instance (IChannel/User).
+        /// </summary>
+        /// <param name="globalListeners">Global listeners of T. Given listener will be added to this.</param>
+        /// <param name="listeners">Currently active Listeners of T. Given listener will be added to this.</param>
+        /// <param name="mapper">Mapper from T -> T' ⊂ T.</param>
+        /// <param name="listener">Listener to be called if T' changes.</param>
+        /// <typeparam name="T">IReadModifyWriteObject-Type to register listener on.</typeparam>
+        private void AddGlobalListener<T>(ConcurrentBag<GlobalListener<T>> globalListeners,
+                                          ConcurrentDictionary<int, ImmutableList<ActiveListener<T>>> listeners,
+                                          Func<T, object> mapper, Func<T, Task> listener) where T : IReadModifyWriteObject
+        {
+            var gl = new GlobalListener<T>(mapper, listener);
+            globalListeners.Add(gl);
+            foreach (var key in listeners.Keys)
```

Comment:
More like a general comment: You're relying on the `listeners` dict having a key for every online user. Which is why you set them up this way in `Init()`. As a consequence, the `AddOrUpdate()` should never add and always update (bug 1: if someone disconnects while you're adding a listener, they will first get removed from the dict and then re-added).

And you'll not only get into tons of awkwardness when abusing APIs. Stuff will break. Your code, usually.

Bug 2: If someone connects after your `Init` and the plugin hasn't added listeners (yet), they won't be added to the list.

So in the end: Please just use things they way they're meant to be used. If you want a list, just use a list. And *not* `Dictionary.Keys`!